monashspa/PHS3302/d0_lifetime/d0_lifetime.py

```diff
@@ -0,0 +1,410 @@
+# These two different contributions will be easy to see if you've calculated the $D^0$ mass correctly. Hint: the mass of the $D^0$ meson is $1864.84\pm0.05\,{\rm MeV}/c^2$. 
+# 
+# **Task 3:** *[2 marks]* Create a histogram of the $D^0$ mass.
+
+
+
+## Plot D0_M histogram
+
+
+# ### Part 2b: $D^0$ lifetime
+# To calculate the lifetime we first need to determine the distance between the proton-proton interaction vertex ($D^0$ production vertex, or primary vertex) and the $D^0$ decay vertex (or secondary vertex). The 3-vector between these points is given by:
+# $$
+# \Delta\vec{x} = \vec{x}_{\rm decay} - \vec{x}_{\rm production}.
+# $$
+# Note, both the production and decay positions are given in units of $\rm mm$. 
+# 
+# **Task 4:** *[4 marks]* Calculate and plot a histgram of the absolute distance $|\Delta\vec{x}|$ between the primary vertex (proton-proton collision) and secondary vertex ($D^0$ decay).
+
+
+
+# D0_L = 
+
+
+
+
+## Plot D0_L histogram
+
+
+# In the lab frame, a particle travelling at velocity $\vec{\mathbf{v}}$ for a time $t$ will travel a distance $\Delta \vec{\mathbf{x}}$:
+# $$
+# \Delta\vec{\mathbf{x}} = \vec{\mathbf{v}}t
+# $$
+# 
+# We must account for special relativity when trying to determine the lifetime of the $D^0$ meson. We can rewrite this equation using relations between the momentum and velocity, and between lab frame time and proper time:
+# 
+# $$
+# \vec{\mathbf{p}} = \gamma m \vec{\mathbf{v}}\\
+# t = \gamma \tau.
+# $$
+# 
+# From this we find: 
+# $$
+# \Delta\vec{\mathbf{x}} = \frac{\vec{\mathbf{p}}}{\gamma m} \gamma \tau = \tau\frac{ \vec{ \mathbf{p}}}{m}
+# $$
+# Multiplying both sides by $\vec{\mathbf{p}}$ and rearranging gives
+# $$
+#  \tau  = \frac{m}{|p^2|} \vec{\mathbf{p}}.\Delta\vec{\mathbf{x}}.
+# $$
+# 
+# If we use $\vec{\mathbf{p}}$ in units of $[{\rm MeV}]$, $m$ in units of $[{\rm MeV}]$ and $\Delta\vec{\mathbf{x}}$ in units of $[\rm mm]$, the proper time $\tau$ will be calculated in units of $[\rm mm]$ (i.e. it's actually a length. This is because we are using natural units for the momentum and mass but SI units for the positions). 
+# 
+# We can convert $[\rm mm]$ into $[\rm ps]$ by dividing  by the speed of light in units of $[\rm mm \,ps^{-1}]$ (given at the top of the script).
+
+# *Aside:* this method is of determining the lifetime is only approxmate. This is because the particles travel through a magnetic field which bends their trajectory, allowing us to measure their momentum. As a result, the momentum 3-vectors of the kaons and pions point in a slightly different direction depending on where we are measuring them. However, the momenta we used to find the vertex assumed the kaon and pion originated at the proton-proton interaction, which we don't expect to be true! Therefore finding the best estimate of the decay time requires iteratively changing the decay position and recalculating the momentum until the values converge.    
```

Comment:
Maybe just delete this. As we do not use it, it wil just cause confusion.

Reply:
I've taken out the aside